### PR TITLE
Basic sampling thread for JFR

### PR DESCRIPTION
**Pull request by gacholio**

### CODE REVIEW

**Comment by gacholio** [comment]
Looks like there may be some memory corruption going on - I'm going to see if the problem is in the event framework or the output writer.

**Comment by gacholio** [comment]
Looks like it's the writer - with that disabled I get a clean build with JFR forced on for every job.

With the writer enabled, I saw two failures:

```
    [javac] Compiling 66 source files to /tmp/bld_75494/workspace/test/functional/DDR_Test/bin
    [javac] JVMJ9VM249W Either the OPENJ9_METADATA_BLOB_FILE_PATH environment variable is not set, or the file does not exist in the specified location.
    [javac] javac: malloc.c:3727: _int_malloc: Assertion `(unsigned long) (size) >= (unsigned long) (nb)' failed.
```

which at that point hangs. The other is:

```
    [javac] Unhandled exception
    [javac] Type=Segmentation error vmState=0x00000000
    [javac] J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
    [javac] Handler1=00007F5CB3D57FF0 Handler2=00007F5CB3C2F7B0 InaccessibleAddress=0000000000000000
    [javac] RDI=00007F5C60007D80 RSI=00007F5CA967A5C0 RAX=0000000000000078 RBX=00007F5C60017840
    [javac] RCX=0000000000000125 RDX=0000000000000000 R8=00007F5C60007E20 R9=00007F5CAC8E89D4
    [javac] R10=00007F5CAC8E4FC8 R11=00007F5CAC8E3290 R12=00007F5CB3E57DC0 R13=0000000000000000
    [javac] R14=00007F5C60007E10 R15=00007F5CAC00EF00
    [javac] RIP=00007F5CB3CCCCDA GS=0000 FS=0000 RSP=00007F5CA967A520
    [javac] EFlags=0000000000010246 CS=0033 RBP=00007F5C60007D80 ERR=0000000000000004
    [javac] TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000000
    [javac] xmm0=0000010400000000 (f: 0.000000, d: 5.517189e-312)
    [javac] xmm1=0000000000000104 (f: 260.000000, d: 1.284571e-321)
    [javac] xmm2=00000000b7654321 (f: 3076866816.000000, d: 1.520174e-314)
    [javac] xmm3=0000000000000002 (f: 2.000000, d: 9.881313e-324)
    [javac] xmm4=0000000000016800 (f: 92160.000000, d: 4.553309e-319)
    [javac] xmm5=000000000000604e (f: 24654.000000, d: 1.218069e-319)
    [javac] xmm6=000000060000002d (f: 45.000000, d: 1.273197e-313)
    [javac] xmm7=0000000000000000 (f: 0.000000, d: 0.000000e+00)
    [javac] xmm8=0000000000000000 (f: 0.000000, d: 0.000000e+00)
    [javac] xmm9=ffffffffffffffff (f: 4294967296.000000, d: -nan)
    [javac] xmm10=0000000000000000 (f: 0.000000, d: 0.000000e+00)
    [javac] xmm11=ffffffffffffffff (f: 4294967296.000000, d: -nan)
    [javac] xmm12=0000000000000000 (f: 0.000000, d: 0.000000e+00)
    [javac] xmm13=0022b20212000d9f (f: 301993376.000000, d: 5.199815e-308)
    [javac] xmm14=0048b613120045b4 (f: 302007744.000000, d: 2.749215e-307)
    [javac] xmm15=124c000ba70049b8 (f: 2801814016.000000, d: 1.549223e-220)
    [javac] Module=/bluebird/builds/bld_75481/sdk/xa6480/jre/lib/amd64/compressedrefs/libj9vm29.so
    [javac] Module_base_address=00007F5CB3CBA000
    [javac] Target=2_90_20240731_75481 (Linux 4.12.14-122.222-default)
    [javac] CPU=amd64 (4 logical CPUs) (0x1e341c000 RAM)
    [javac] ----------- Stack Backtrace -----------
    [javac] pool_newElement.cold+0x0 (0x00007F5CB3CCCCDA [libj9vm29.so+0x12cda])
    [javac] hashTableAddNodeInList+0x9c (0x00007F5CB3E040FC [libj9vm29.so+0x14a0fc])
    [javac] _ZN23VM_JFRConstantPoolTypes18addStringUTF8EntryEP6J9UTF8b+0xa3 (0x00007F5CB3D23FC3 [libj9vm29.so+0x69fc3])
    [javac] _ZN23VM_JFRConstantPoolTypes13getClassEntryEP7J9Class+0x6d (0x00007F5CB3D244AD [libj9vm29.so+0x6a4ad])
    [javac] _ZN23VM_JFRConstantPoolTypes14getMethodEntryEP11J9ROMMethodP7J9Class+0x5b (0x00007F5CB3D2463B [libj9vm29.so+0x6a63b])
    [javac] _ZN23VM_JFRConstantPoolTypes18stackTraceCallbackEP10J9VMThreadPvmP10J9ROMClassP11J9ROMMethodP6J9UTF8mP13J9ClassLoaderP7J9Class+0x46 (0x00007F5CB3D25546 [libj9vm29.so+0x6b546])
    [javac] iterateStackTraceImpl+0x419 (0x00007F5CB3D50359 [libj9vm29.so+0x96359])
    [javac] _ZN23VM_JFRConstantPoolTypes23addExecutionSampleEntryEP20J9JFRExecutionSample+0xbd (0x00007F5CB3D24D0D [libj9vm29.so+0x6ad0d])
    [javac] _ZN12VM_JFRWriter18flushJFRDataToFileEP10J9VMThreadb.isra.0+0x209 (0x00007F5CB3D5AB19 [libj9vm29.so+0xa0b19])
    [javac] jfrThreadDestroy+0x65 (0x00007F5CB3D5B695 [libj9vm29.so+0xa1695])
    [javac] J9HookDispatch+0x136 (0x00007F5CB3C9A966 [libj9hookable29.so+0x1966])
    [javac] deallocateVMThread+0x32f (0x00007F5CB3D6A66F [libj9vm29.so+0xb066f])
    [javac] javaProtectedThreadProc+0xbf (0x00007F5CB3D94F7F [libj9vm29.so+0xdaf7f])
    [javac] omrsig_protect+0x239 (0x00007F5CB3C30439 [libj9prt29.so+0x27439])
    [javac] javaThreadProc+0x49 (0x00007F5CB3D91859 [libj9vm29.so+0xd7859])
    [javac] thread_wrapper+0x163 (0x00007F5CB3CAC393 [libj9thr29.so+0xb393])
    [javac] start_thread+0xca (0x00007F5CB4D8271A [libpthread.so.0+0x971a])
    [javac] clone+0x6d (0x00007F5CB4A9C21D [libc.so.6+0xf021d])
```

File: runtime/vm/jfr.cpp [hunk 1]
```diff
@@ -602,15 +662,56 @@ initializeEventFields(J9VMThread *currentThread, J9JFREvent *event, UDATA eventT
...
+		omrthread_monitor_enter(vm->jfrSamplerMutex);
+		vm->jfrSamplerState = J9JFR_SAMPLER_STATE_RUNNING;
+		omrthread_monitor_notify_all(vm->jfrSamplerMutex);
+		while (J9JFR_SAMPLER_STATE_STOP != vm->jfrSamplerState) {
+			omrthread_monitor_exit(vm->jfrSamplerMutex);
+			internalEnterVMFromJNI(currentThread);
+			acquireExclusiveVMAccess(currentThread);
+			J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
+			while (NULL != walkThread) {
+				jfrExecutionSample(currentThread, walkThread);
```
Comment (tajila) [inline runtime/vm/jfr.cpp:696]: im seeing a lot of empty stack traces, perhaps we should wrap this with

```
				if (VM_VMHelpers::threadCanRunJavaCode(walkThread)) {
					jfrExecutionSample(currentThread, walkThread);
				}
```

Comment (gacholio) [reply runtime/vm/jfr.cpp]: Done.

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk17

**Comment by gacholio** [comment]
jenkins test sanity xlinux jdk17

**Pull request merged**